tests/Bitcoin/TxComparisonHelper.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright © 2017-2020 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "TxComparisonHelper.h"
+
+#include <TrustWalletCore/TWCoinType.h>
+
+#include "Bitcoin/OutPoint.h"
+#include "proto/Bitcoin.pb.h"
+#include "Data.h"
+
+#include <gtest/gtest.h>
+#include <sstream>
+#include <cassert>
+
+using namespace TW;
+using namespace TW::Bitcoin;
+
+auto emptyTxOutPoint = OutPoint(Data(32), 0);
+
+Proto::UnspentTransaction buildTestUTXO(int64_t amount) {
+    Proto::UnspentTransaction utxo;
+    utxo.set_amount(amount);
+    const auto& outPoint = emptyTxOutPoint;
+    utxo.mutable_out_point()->set_hash(outPoint.hash.data(), outPoint.hash.size());
+    utxo.mutable_out_point()->set_index(outPoint.index);
+    return utxo;
+}
+
+std::vector<Proto::UnspentTransaction> buildTestUTXOs(const std::vector<int64_t>& amounts) {
+    std::vector<Proto::UnspentTransaction> utxos;
+    for (auto it = amounts.begin(); it != amounts.end(); it++) {
+        utxos.push_back(buildTestUTXO(*it));
+    }
+    return utxos;
+}
+
+Proto::SigningInput buildSigningInput(Amount amount, int byteFee, const std::vector<Proto::UnspentTransaction>& utxos, bool useMaxAmount, enum TWCoinType coin) {
+    Proto::SigningInput input;
+    input.set_amount(amount);
+    input.set_byte_fee(byteFee);
+    input.set_use_max_amount(useMaxAmount);
+    input.set_coin_type(coin);
+    *input.mutable_utxo() = { utxos.begin(), utxos.end() };
+    return input;
+}
+
+int64_t sumUTXOs(const std::vector<Proto::UnspentTransaction>& utxos) {
+    int64_t s = 0u;
+    for (auto& utxo: utxos) {
+        s += utxo.amount();
+    }
+    return s;
+}
+
+std::string verifySelectedUTXOs(const std::vector<Proto::UnspentTransaction>& selected, const std::vector<int64_t>& expectedAmounts) {
+    std::stringstream ss;
+    if (selected.size() != expectedAmounts.size()) {
+        ss << "Wrong number of selected UTXOs, " << selected.size() << " vs. " << expectedAmounts.size() << std::endl;
+    }
+    for (auto i = 0; i < selected.size() && i < expectedAmounts.size(); ++i) {
+        if (expectedAmounts[i] != selected[i].amount()) {
+            ss << "Wrong UTXOs amount, pos " << i << " amount " << selected[i].amount() << " expected " << expectedAmounts[i] << std::endl;
+        }
+    }
+    return ss.str();
+}
+
+std::string verifyPlan(const TransactionPlan& plan, const std::vector<int64_t>& utxoAmounts, int64_t outputAmount, int64_t fee) {
```

Comment:
is there a better way to handle errors? it strange to return `string`

Reply:
Well, gtest ASSERTs are valid only within testcases (TEST() {...}), not outside.  The test context could be passed to helper function, but that would require tricks to dig out the test context hidden in the TEST macro -- ugly (I hate macros!).
Another option is to simple print differences on stderr, and only return bool for assert.  I found it more informative to return the error message as string.  But this is test only, of course.
Any better suggestions, anyone?

Reply:
Return `bool` sounds more understanding to me, and use logging for printing an error.

Reply:
It is a bit cleaner, right